src/uint/modular/macros.rs

```diff
@@ -0,0 +1,39 @@
+// TODO: Use `adt_const_params` once stabilized to make a `Residue` generic around a modulus rather than having to implement a ZST + trait
+#[macro_export]
+/// Implements a modulus with the given name, type, and value, in that specific order. Please `use crypto_bigint::traits::Encoding` to make this work.
+/// For example, `impl_modulus!(MyModulus, U256, "73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001");` implements a 256-bit modulus named `MyModulus`.
+macro_rules! impl_modulus {
+    ($name:ident, $uint_type:ty, $value:expr) => {
+        #[derive(Clone, Copy, PartialEq, Eq, Debug)]
+        pub struct $name {}
+        impl<const DLIMBS: usize> ResidueParams<{ nlimbs!(<$uint_type>::BIT_SIZE) }> for $name
+        where
+            $crate::UInt<{ nlimbs!(<$uint_type>::BIT_SIZE) }>:
+                $crate::traits::Concat<Output = $crate::UInt<DLIMBS>>,
+            $crate::UInt<DLIMBS>: $crate::traits::Split<Output = $uint_type>,
+        {
+            const LIMBS: usize = { nlimbs!(<$uint_type>::BIT_SIZE) };
+            const MODULUS: $crate::UInt<{ nlimbs!(<$uint_type>::BIT_SIZE) }> =
+                <$uint_type>::from_be_hex($value);
+            const R: $crate::UInt<{ nlimbs!(<$uint_type>::BIT_SIZE) }> = $crate::UInt::MAX
+                .ct_reduce(&Self::MODULUS)
+                .0
+                .wrapping_add(&$crate::UInt::ONE);
+            const R2: $crate::UInt<{ nlimbs!(<$uint_type>::BIT_SIZE) }> =
+                $crate::UInt::ct_reduce_wide(Self::R.square_wide(), &Self::MODULUS).0;
```

Comment:
If this is the only usage of `square_wide`, it seems like it can be trivially replaced with `Self::R.mul_wide(&Self::R)`

Reply:
Is it okay if I leave it like this so I can implement a faster square method later today? Modular exponentiation would benefit greatly from an optimized squaring method. 😄